src/auth/authenticate.rs

```diff
@@ -0,0 +1,275 @@
+}
+
+/// Builds a `WWW-Authenticate` header according to the following scheme:
+///
+/// `PrivateToken challenge=... token-key=... [max-age=...]`
+pub fn build_www_authenticate_header(
+    token_challenge: TokenChallenge,
+    token_key: &[u8],
+    max_age: Option<usize>,
+) -> Result<String, BuildError> {
+    let challenge_value = token_challenge
+        .to_base64()
+        .map_err(|_| BuildError::InvalidTokenChallenge)?;
+    let token_key_value = base64::encode(token_key);
+    let max_age_string = if let Some(max_age) = max_age {
+        format!(", max-age={}", max_age)
+    } else {
+        "".to_string()
+    };
+
+    let value = format!(
+        "PrivateToken challenge={}, token-key={}{}",
+        challenge_value, token_key_value, max_age_string
+    );
+    Ok(value)
+}
+
+/// Building error for the `Authorization` header values
+#[derive(Error, Debug)]
+pub enum BuildError {
+    #[error("Invalid TokenChallenge")]
+    InvalidTokenChallenge,
+}
+
+/// Parses a `WWW-Authenticate` header according to the following scheme:
+///
+/// `PrivateToken challenge=... token-key=... [max-age=...]`
+pub fn parse_www_authenticate_header(value: &str) -> Result<Vec<Challenge>, ParseError> {
+    WwwAuthenticateParser::try_from_str(value)
+}
+
+/// Decoded challenge from a `WWW-Authenicate` header
+#[derive(Debug, PartialEq)]
+pub struct Challenge {
+    challenge: String,
+    token_key: Vec<u8>,
+    max_age: Option<usize>,
+}
+
+/// Parsing error for the `WWW-Authenticate` header values
+#[derive(Error, Debug)]
+pub enum ParseError {
+    #[error("Invalid challenge")]
+    InvalidChallenge,
+    #[error("Invalid token key")]
+    InvalidTokenKey,
+    #[error("Invalid max age")]
+    InvalidMaxAge,
+    #[error("Invalid input string")]
+    InvalidInput,
+}
+
+#[derive(Parser)]
+#[grammar_inline = r#"
+WHITESPACE = _{ " " }
+name_char = { ASCII_ALPHANUMERIC | "-" }
+base64_char = { ASCII_ALPHANUMERIC | "+" | "/" | "=" }
+name = @{ name_char+ }
+value = @{ base64_char* }
+num_value = { ASCII_DIGIT+ }
+property = { name ~ "=" ~ value }
+challenge_param = { "challenge=" ~ value }
+token_key_param = { "token-key=" ~ value }
+max_age_param = { "max-age=" ~ num_value }
+challenge = { "PrivateToken " ~ challenge_param ~ "," ~ token_key_param ~ ("," ~ max_age_param)? }
+challenge_list = {
+    SOI ~
+     ((challenge ~ ",") | challenge)+ ~
+    EOI
+}
+"#]
+struct WwwAuthenticateParser {}
+
+impl WwwAuthenticateParser {
+    fn try_from_str(value: &str) -> Result<Vec<Challenge>, ParseError> {
+        let mut challenges = Vec::new();
+        let challenge_list = Self::parse(Rule::challenge_list, value)
+            .map_err(|_| ParseError::InvalidInput)?
+            .next()
+            .ok_or(ParseError::InvalidInput)?
+            .into_inner();
+        for challenge in challenge_list {
+            let mut params = challenge.into_inner();
+
+            if params.peek().is_some() {
+                let challenge_param = params
```

Comment:
Does this assume a specific order for the parameters? It should be possible for the token-key parameter to come before the challenge parameter.

Reply:
You are correct, the order of parameters is fixed right now. I started with something more flexible, but then went back to an MVP.

The parser definitely needs more work:

 - Order of parameters should be flexible
 - Unknown parameters should be ignored/ or reported back to the consumer
 - Are keywords like "PrivateToken" and parameter names case-sensitive? (right now they are)
 - Are whitespace around the "=" allowed (right now they aren't)

Reply:
OK, no problem. Let's track this in a followup PR. Baby steps. =)